torch/_inductor/ir.py

```diff
@@ -4015,3 +4015,106 @@ def debug_str(self, name="block"):
+        # perspective that Wait's buffer is the one being consumed by consumers, to ensure
+        # wait op gets scheduled at all.  And as a side effect, wait's buffer name needs to actually
+        # point to something in the codegen output
+
+        # this 'del' is screwing up scheduler 'reuse' of this buf. which i'm not sure should even be allowed
+        # wrapper.writeline(f"{self.get_name()} = {all_reduce}; del {all_reduce}")
+        wrapper.writeline(f"{self.get_name()} = {all_reduce}")
+
+
+class AllReduce(ExternKernel):
+    def __init__(
+        self,
+        layout,
+        inputs,
+        constant_args=(),
+    ):
+        super().__init__(None, layout, inputs, constant_args)
+        self.name = V.graph.register_buffer(self)
+
+    def should_allocate(self):
+        return True
+
+    @classmethod
+    def create(
+        cls,
+        x: "TensorBox",
+        group_id: int,
+        reduce_op: str,
+    ):
+        x = cls.realize_input(x)
+
+        # is there a difference between literally using x.data.layout below, vs
+        # creating a new one that has the same properties?
+        new_layout = FlexibleLayout(x.get_device(), x.get_dtype(), x.get_size())
+
+        # AllReduce returns a 'work' object.  But Inductor's scheduler doesn't need to know
+        # about that, and we just pretend for scheduling purposes that the work obj is a 1-elem tensor.
+        # Nobody should consume the output of AllReduce except 'Wait', which we control here.
+        all_reduce = AllReduce(
+            layout=new_layout,
+            inputs=[x],
+            constant_args=[group_id, reduce_op],
+        )
+
+        # Return a 'Wait' to the user that called 'all_reduce' in the first place.  It consumes the 'work'
+        # and waits on it, also producing a buffer which is really the input buffer to AllReduce.
+        return Wait(
+            layout=new_layout,
+            inputs=[all_reduce],
+        )
+
+    def codegen(self, wrapper):
+        # these header lines just get output once
+        wrapper.header.writeline("import torch.distributed as dist")
+        wrapper.header.writeline("from torch._C._distributed_c10d import ReduceOp")
+
+        # extract references to our args in string form for codegen output
+        (input_name,) = [t.codegen_reference() for t in self.inputs]
+        output_name = self.get_name()
+        group_id = f"{repr(self.constant_args[0])}"
+        reduce_op = self.constant_args[1]
+        # TODO make this real
+        c10d_op = {"sum": "ReduceOp.SUM"}
+
+        # We must copy our input buffer sometimes, but the scheduler will help us find opportunities
+        # to reuse the input buffer.  (This requires no other users of the input buffer.)
+        sched_node = V.graph.scheduler.name_to_node[output_name]
+        input_read = next(iter(sched_node.read_writes.reads))
+        if not sched_node.can_inplace(input_read):
```

Comment:
Double check to make sure this is robust.  Not totally convinced about correctness.

Reply:
Might be better to move this flag to into the caller that might have more information.

Reply:
by caller, you mean lowering.py? and which information would I be looking for?

I think my biggest issue with this is that "can_inplace" does not imply "did_inplace".  So I considered actually setting up a new map of 'things_inplaced" inside the scheduler that I could peek at here.  And perhaps if I set up the mutation thing correctly, I could just reuse the mutation map for this purpose (although I am not sure if that would lead to a false positive in this logic in some cases).

Reply:
Oh I was thinking of trying to get this in wrapper.py (where allocations happen)